product_tier_validation/models/product_template.py

```diff
@@ -8,3 +8,34 @@ class ProductTemplate(models.Model):
     _name = "product.template"
     _inherit = ["product.template", "tier.validation"]
     _tier_validation_manual_config = False
+
+    def _check_state_conditions(self, vals):
+        """
+        As product state write is done on product_state_id field if
+        changed through interface. We need to get the correspondong state
+        """
+        self.ensure_one()
+        if "product_state_id" in vals:
+            new_state = self.env["product.state"].browse(vals["product_state_id"]).code
+            return (
+                getattr(self, self._state_field) in self._state_from
+                and new_state in self._state_to
+            )
+        else:
+            return super()._check_state_conditions(vals)
+
+    def _check_tier_state_transition(self, vals):
+        """
+        As product state write is done on product_state_id field if
+        changed through interface. We need to get the correspondong state
+        """
+        self.ensure_one()
+        if "product_state_id" in vals:
+            new_state = self.env["product.state"].browse(vals["product_state_id"]).code
+            return getattr(
+                self, self._state_field
+            ) in self._state_from and new_state not in (
+                self._state_to + [self._cancel_state]
+            )
+        else:
+            return super()._check_tier_state_transition(vals)
```

Comment:
With the improvement at https://github.com/hailangvn/server-ux/pull/1, we can rewrite this like below.
```python
    def _get_state_from(self, vals):
        return self.env["product.state"].browse(vals["product_state_id"]).code
```

Reply:
wouldn't be better something like:

```py
[...]
state_code = fields.Char(related="product_state_id.code")`

_state_field = "state_code"
_state_from = ["draft"]
_state_to = ["sellable"]
[...]
```

and we don't override anything?